JS Constructor Attr Support

ambiguousname opened the pull request:
Fixes #718.

Current TODOs:

- [x] Add consistent functions that can always be called (`fromFields`, `fromValue`, etc.) for when the default constructor is overridden.
- [x] Hide private functions from .d.ts files
- [x] Add test for overriding constructor for Enums (and calling `fromValue`)
- [x] Ensure passing tests
- [x] Update demo_gen to match
- [x] Fix whitespacing
- [x] See if `usage` is removable (I think this can be avoided by just `return`ing a new constructed object)
- [x] Avoid using symbol parameters (and `...args`) in `constructor` functions so those inspecting source can see what the default constructor takes for params

ambiguousname commented:
This got a little messy, but I think the rough roadwork is here.

To sum the changes up (in addition to allowing overriding constructor functions within Diplomat):
- I added a `usage` parameter to `gen_c_to_js_for_type` to specifically check if we are looking to generate a type from scratch or inside of a constructor function.
- I created a bunch of new helper methods to make construction easier (`_fromFFI` has changed a bunch and has new helpers, specifically).
- I added a new `js_class.js.jinja` file that Enums, Structs, and Opaques are now all based on.
- I added a new `exposeConstructor` symbol specifically to access internal constructors that are overridden by default constructors (although new constructors also make use of `internalConstructor`, see the `js_class.js.jinja` comments for more on this)

Manishearth commented:
Let me know when this is ready to be reviewed!

ambiguousname commented:
Need to update demo_gen to match new JS outputs, but everything else is ready for review

ambiguousname commented:
The dart tests failing seems like it's from the fact that I used

`#[diplomat::attr(auto, constructor)]` on an Enum.

Is this something our Dart implementation supports? Should I add a feature gate or something like that?

Manishearth commented:
Looks like Dart supports named factory constructors on enums, but not default ones.

For now I'm fine with having this be `all(supports = constructors, not(dart))`

Manishearth commented on tool/templates/js/enum.js.jinja:
```diff
@@ -16,10 +13,8 @@ export class {{type_name}} {
-    but we just need to check if a constructor will be generated. -#}
-    constructor(value) {
+    
+    {% if overrides_constructor -%} #internalConstructor {%- else -%} constructor {%- endif -%}(value) {
```
issue: I would unconditionally call this `internalConstructor`. constructors work differently from methods, we should not use the same code for both

All branches of this method should `return this` OR return an object.

Manishearth commented on tool/templates/js/runtime.mjs:
```diff
@@ -1,5 +1,7 @@
 /** For internal Diplomat use when constructing opaques or structs. */
 export const internalConstructor = Symbol("constructor");
+/** For internal Diplomat use when accessing an internal constructor.  */
+export const exposeConstructor = Symbol("exposeConstructor");
```
nit: let's document the differences between the two Constructor symbols

Manishearth reviewed (commented):
Overall this seems good but I really don't think we should ever be sharing codegen between constructors and regular JS methods: a codegenned method should be one or the other.


The reason is that `constructor` has some subtle behavior: in a `constructor`, you either mutate `this`, or you return an object: If what you return is an Object (_any_ object), that is the construction result.

We need to go through a constructor eventually since JS does not provide a good way to make objects with the right `typeof` without using a `constructor`.

What we should have is:

 - `internalConstructor`: A non-static method that takes `this` and returns an object. This method should mutate `this` if necessary, _and_ always return `this` or another object of the right type:
   - For opaques:  set up lifetimes, and set up `this.ptr`. Return `this`
   - For structs: set subfields. Return `this`.
   - For enums: either return a cached object, the provided object, or (internal-only) set `this.value` and return `this`
 - `constructor`: A thin wrapper that returns `this.internalConstructor()`, or a custom ctor.
 - `fromFields`/`fromValue`: Thin wrappers around `new Foo(exposeConstructor?, ...)`
 - `fromFfi`: We don't have this anymore, but this is just `new Foo(internalConstructor, ...)`

I don't think returning `this` is strictly always necessary: for example the struct setup works fine without it, but it might be better from a hygeine point of view? Open to other ideas. The nice thing about this design is that the codegenned constructor can also be identical in that case.

Thoughts?

ambiguousname commented:
I think all of your suggestions sound good to avoid future headaches.

I believe what you're describing for `#internalConstructor` ONLY mutating `this` is very similar to what constructors looked like when I needed a `usage : Option<SpecialMethod>` parameter in `gen_c_to_js_for_type` to determine whether we were interested in modifying `this` or returning a new object (based on whether or not `usage` was `Some(SpecialMethod::Constructor)` or not)

If you're curious as to what that looked like, the commit hash was `a70e7005d2b3f295926b4f641efb96a9bdbf3f7a`

When I have some time, I'll go about changing all `#internalConstructor` methods to only modify `this`, instead of returning a new object.

ambiguousname commented:
On further reading though, I think I may be missing your point. My understanding now is that you'd like to completely remove any form of sharing generation logic for any method that is meant to be a default constructor. So we might move all that logic from `generate_method` to an entirely new function, like say `generate_constructor_method`? Do I have that right?

ambiguousname commented:
Ah, I believe I understand your point now. I'll start by moving the generation for `constructor` into one place (`js_class.js.jinja`) and work from there.

ambiguousname commented:
@Manishearth I moved logic for generating `constructor`s to `js_class.js.jinja` (and as a consequence of that, moved logic  from `special_methods.js.jinja` to `js_class`).

Let me know if this matches what you had in mind!

Manishearth commented on tool/src/js/gen.rs:
```diff
@@ -413,7 +424,11 @@ impl<'tcx> TyGenContext<'_, 'tcx> {
+
+            /// Used by `js_class.js.jinja`. If a constructor isn't overridden by #[diplomat::attr(auto, constructor)], this is the logic that `js_class.js.jinja` will use to determine whether or not to generate constructor code.
+            /// Useful for hiding the fact that an out_struct has a constructor in typescript headers, for instance.
+            show_default_ctor: bool,
```
thought (followup): we should probably move shared types into a shared ImplTemplate type, and have specific StructImplTemplate/etc

Manishearth commented on tool/templates/js/runtime.mjs:
```diff
@@ -1,5 +1,11 @@
+ * This is for when we're handling items that we don't want the user to touch, like an structure that's only meant to be output, or de-referencing a pointer we're handed from WASM.
+ */
 export const internalConstructor = Symbol("constructor");
+/** For internal Diplomat use when accessing an internal constructor that's been overridden by a default constructor.
```
```suggestion
/** For internal Diplomat use when accessing a from-fields/from-value constructor that's been overridden by a default constructor.
```

Manishearth reviewed (approved):
praise: nice work!!
